lib/weighted_resblock.py

```diff
@@ -0,0 +1,321 @@
+    self.num_templates = num_templates
+    self.kernel_reg = kernel_reg
+
+  def build(self, input_shape):
+    input_channel = input_shape[0][-1]
+    self.expanded_channel = input_channel*self.expansion_factor
+
+    kernel_init, bias_init = self.template.get_expansion_template()
+    self.conv1 = wl.WeightedConv2D(
+        filters=self.expanded_channel, kernel_size=1, strides=(1, 1),
+        padding='same', num_templates=self.num_templates,
+        kernel_initializer=kernel_init,
+        kernel_regularizer=regularizers.l2(self.kernel_reg),
+        bias_initializer=bias_init)
+    self.conv1.build(input_shape)
+
+    gamma, beta, mean, var = self.template.get_bn1_template()
+    self.bn1 = wl.WeightedBatchNormalizationSeparate(
+        num_templates=self.num_templates, gamma_initializer=gamma,
+        beta_initializer=beta, moving_mean_initializer=mean,
+        moving_variance_initializer=var)
+
+    depthwise_kernel_init, bias_init = self.template.get_depthwise_template()
+    self.conv2 = wl.WeightedDepthwiseConv2D(
+        kernel_size=self.kernel_size, strides=(1, 1), padding='same',
+        num_templates=self.num_templates,
+        depthwise_initializer=depthwise_kernel_init,
+        #depthwise_regularizer=regularizers.l2(1e-4),
+        bias_initializer=bias_init)
+
+    cov2_in_shape = ((input_shape[0][0], input_shape[0][1], input_shape[0][2],
+                      self.expanded_channel), (self.num_templates,))
+    self.conv2.build(cov2_in_shape)
+
+    gamma, beta, mean, var = self.template.get_bn2_template()
+    self.bn2 = wl.WeightedBatchNormalizationSeparate(
+        num_templates=self.num_templates, gamma_initializer=gamma,
+        beta_initializer=beta, moving_mean_initializer=mean,
+        moving_variance_initializer=var)
+
+    kernel_init, bias_init = self.template.get_projection_template()
+    self.conv3 = wl.WeightedConv2D(
+        filters=input_channel, kernel_size=1, strides=(1, 1), padding='same',
+        num_templates=self.num_templates,
+        kernel_initializer=kernel_init,
+        kernel_regularizer=regularizers.l2(self.kernel_reg),
+        bias_initializer=bias_init)
+    self.conv3.build(cov2_in_shape)
+
+    gamma, beta, mean, var = self.template.get_bn3_template()
+    self.bn3 = wl.WeightedBatchNormalizationSeparate(
+        num_templates=self.num_templates, gamma_initializer=gamma,
+        beta_initializer=beta, moving_mean_initializer=mean,
+        moving_variance_initializer=var)
+
+    self.built = True
+
+  def call(self, inputs, training=None):
+    layer_input, mix_weights = inputs
+    x = self.conv1([layer_input, mix_weights])
+    x = self.bn1([x, mix_weights], training)
+    x = self.activation(x)
+
+    x = self.conv2([x, mix_weights])
+    x = self.bn2([x, mix_weights], training)
+    x = self.activation(x)
+
+    x = self.conv3([x, mix_weights])
+    x = self.bn3([x, mix_weights], training)
+    x += layer_input
+    return x
+
+  def set_activation_fn(self, activation):
```

Comment:
I would call it differently. It does not really set the activation function, but maps the name of the activation to the actual function.